File: canonical-collaboration-demo/server/app/jobs.py
import asyncio
from datetime import timedelta
import json
from uuid import uuid4

from fastapi import HTTPException

from .agent import DocumentAgent
from .config import settings
from .models import CreateJobRequest, JobRecord, JobStatus
from .rooms import RoomStore, now


class JobService:
    def __init__(self, rooms: RoomStore, agent: DocumentAgent) -> None:
        self._rooms = rooms
        self._agent = agent
        self._queue: asyncio.Queue[str] = asyncio.Queue()
        self._records: dict[str, JobRecord] = {}
        self._running: dict[str, asyncio.Task[str]] = {}
        self._worker_task: asyncio.Task | None = None

    def start(self) -> None:
        self._worker_task = asyncio.create_task(self._work())

    async def stop(self) -> None:
        if self._worker_task:
            self._worker_task.cancel()
            await asyncio.gather(self._worker_task, return_exceptions=True)

    async def create(self, room_id: str, request: CreateJobRequest) -> JobRecord:
        await self._rooms.require(room_id)
        job_id = f"job_{uuid4().hex}"
        timestamp = now()
        record = JobRecord(
            id=job_id,
            room_id=room_id,
            prompt=request.prompt.strip(),
            isSuggesting=request.isSuggesting,
            status=JobStatus.queued,
            created_at=timestamp,
            updated_at=timestamp,
        )
        self._records[job_id] = record
        await self._queue.put(job_id)
        print(json.dumps({"event": "job.queued", "jobId": job_id, "roomId": room_id}), flush=True)
        return record

    def get(self, room_id: str, job_id: str) -> JobRecord:
        record = self._records.get(job_id)
        if not record or record.room_id != room_id:
            raise HTTPException(404, "Job not found.")
        return record

    async def cancel(self, room_id: str, job_id: str) -> JobRecord:
        record = self.get(room_id, job_id)
        if record.status in (JobStatus.completed, JobStatus.failed, JobStatus.cancelled):
            return record
        record.status = JobStatus.cancelled
        record.updated_at = now()
        task = self._running.get(job_id)
        if task:
            task.cancel()
        print(json.dumps({"event": "job.cancelled", "jobId": job_id, "roomId": room_id}), flush=True)
        return record

    def expire(self) -> list[str]:
        cutoff = now() - timedelta(seconds=settings.completed_job_ttl_seconds)
        expired = [
            job_id
            for job_id, record in self._records.items()
            if record.status in (JobStatus.completed, JobStatus.failed, JobStatus.cancelled) and record.updated_at < cutoff
        ]
        for job_id in expired:
            del self._records[job_id]
        return expired

    async def _work(self) -> None:
        while True:
            job_id = await self._queue.get()
            record = self._records.get(job_id)
            if not record:
                self._queue.task_done()
                continue
            if record.status == JobStatus.cancelled:
                self._queue.task_done()
                continue
            room = None
            try:
                room = await self._rooms.require(record.room_id)
                room.active_jobs += 1
                record.status = JobStatus.running
                record.updated_at = now()
                agent_task = asyncio.create_task(
                    self._agent.run(
                        room.editor_document,
                        record.prompt,
                        room.conversation,
                        record.isSuggesting,
                    )
                )
                self._running[job_id] = agent_task
                record.answer = await agent_task
                if record.status == JobStatus.cancelled:
                    continue
                record.status = JobStatus.applying_edit
                record.updated_at = now()
                await asyncio.sleep(0)
                room.last_activity_at = now()
                record.status = JobStatus.completed
            except asyncio.CancelledError:
                if self._worker_task and self._worker_task.cancelling():
                    raise
                record.status = JobStatus.cancelled
            except Exception as error:
                record.status = JobStatus.failed
                record.error = str(error)
            finally:
                record.updated_at = now()
                self._running.pop(job_id, None)
                if room:
                    room.active_jobs = max(0, room.active_jobs - 1)
                print(
                    json.dumps({"event": "job.finished", "jobId": job_id, "status": record.status}),
                    flush=True,
                )
                self._queue.task_done()

File: canonical-collaboration-demo/server/src/jobs.js
import { randomUUID } from 'node:crypto';
import { config } from './config.js';
import { logEvent } from './diagnostics.js';
import { HttpError } from './http-error.js';

const terminalStatuses = new Set(['completed', 'failed', 'cancelled']);

export class JobService {
  #rooms;
  #agent;
  #queue = [];
  #records = new Map();
  #active;
  #running = false;

  constructor(rooms, agent) {
    this.#rooms = rooms;
    this.#agent = agent;
  }

  create(roomId, input) {
    this.#rooms.require(roomId);
    const prompt = typeof input?.prompt === 'string' ? input.prompt.trim() : '';
    if (!prompt || prompt.length > 20_000) throw new HttpError(400, 'Prompt must contain between 1 and 20,000 characters.');
    if (input?.isSuggesting !== undefined && typeof input.isSuggesting !== 'boolean') {
      throw new HttpError(400, 'isSuggesting must be a boolean.');
    }
    const timestamp = new Date().toISOString();
    const record = {
      id: `job_${randomUUID().replaceAll('-', '')}`,
      room_id: roomId,
      prompt,
      isSuggesting: input.isSuggesting ?? true,
      status: 'queued',
      created_at: timestamp,
      updated_at: timestamp,
      answer: null,
      error: null,
    };
    this.#records.set(record.id, record);
    this.#queue.push(record.id);
    logEvent('worker', 'job.queued', { jobId: record.id, roomId, queueSize: this.#queue.length });
    void this.#drain();
    return record;
  }

  get(roomId, jobId) {
    const record = this.#records.get(jobId);
    if (!record || record.room_id !== roomId) throw new HttpError(404, 'Job not found.');
    return record;
  }

  cancel(roomId, jobId) {
    const record = this.get(roomId, jobId);
    if (terminalStatuses.has(record.status)) return record;
    record.status = 'cancelled';
    record.updated_at = new Date().toISOString();
    if (this.#active?.record.id === jobId) this.#active.controller.abort(new Error('Agent job cancelled.'));
    logEvent('worker', 'job.cancelled', { jobId, roomId });
    return record;
  }

  expire() {
    const cutoff = Date.now() - config.completedJobTtlMs;
    const expired = [...this.#records.values()]
      .filter((record) => terminalStatuses.has(record.status) && new Date(record.updated_at).getTime() < cutoff)
      .map((record) => record.id);
    for (const jobId of expired) this.#records.delete(jobId);
    if (expired.length) logEvent('worker', 'jobs.expired', { jobIds: expired, retainedJobs: this.#records.size });
    return expired;
  }

  stop() {
    this.#active?.controller.abort(new Error('Worker stopping.'));
    logEvent('worker', 'worker.stopped');
  }

  async #drain() {
    if (this.#running) return;
    this.#running = true;
    logEvent('worker', 'worker.started');
    try {
      while (this.#queue.length) {
        const jobId = this.#queue.shift();
        const record = this.#records.get(jobId);
        if (!record || record.status === 'cancelled') continue;
        let room;
        const controller = new AbortController();
        this.#active = { record, controller };
        try {
          room = this.#rooms.require(record.room_id);
          room.activeJobs += 1;
          record.status = 'running';
          record.updated_at = new Date().toISOString();
          logEvent('worker', 'job.running', { jobId, roomId: record.room_id, queueSize: this.#queue.length });
          record.answer = await this.#agent.run(
            room.documentId,
            record.prompt,
            room.conversation,
            record.isSuggesting,
            controller.signal,
          );
          if (record.status !== 'cancelled') {
            record.status = 'applying_edit';
            record.updated_at = new Date().toISOString();
            room.lastActivityAt = new Date();
            record.status = 'completed';
          }
        } catch (error) {
          if (controller.signal.aborted || record.status === 'cancelled') {
            record.status = 'cancelled';
          } else {
            record.status = 'failed';
            record.error = error instanceof Error ? error.message : String(error);
            logEvent('worker', 'job.failed', { jobId, roomId: record.room_id, error: record.error });
          }
        } finally {
          record.updated_at = new Date().toISOString();
          if (room) room.activeJobs = Math.max(0, room.activeJobs - 1);
          this.#active = undefined;
          logEvent('worker', 'job.finished', { jobId, roomId: record.room_id, status: record.status, queueSize: this.#queue.length });
        }
      }
    } finally {
      this.#running = false;
    }
  }
}
